packages/protocol/contracts/L1/ProverPool.sol

```diff
@@ -0,0 +1,479 @@
+            totalWeight += weights[i];
+        }
+
+        if (totalWeight == 0) {
+            return (address(0), 0);
+        }
+
+        // Determine prover idx
+        bytes32 rand =
+            keccak256(abi.encode(blockhash(block.number - 1), blockId));
+
+        uint8 id = _pickProver({
+            weights: weights,
+            totalWeight: totalWeight,
+            rand: rand
+        });
+
+        topProvers[id].currentCapacity--;
+        return (idToProver[id], topProvers[id].rewardPerGas);
+    }
+
+    // Increases the capacity of the prover
+    // @Dani: No need to track usedCapacity on prover (exiting), they should take care of
+    // monitoring their own resources when they want to re-enter (!). This code is already
+    // more complex than a PoC should :) - no need to over-engineer.
+    function releaseProver(address prover) external onlyProtocol {
+        /// NOTE: rewardPerGas is the indicator if object is valid/existing
+        if (
+            topProvers[proverToId[prover]].rewardPerGas != 0
+        ) {
+            topProvers[proverToId[prover]].currentCapacity++;
+        }
+    }
+
+    function slashProver(address prover)
+        external
+        onlyProtocol
+        onlyActiveProver(prover)
+    {
+        // We need to determine if the prover is an exiting prover or normal (top)
+        // It might happen that prover is in the exit, while also in the top, in such case
+        // try to punish the exiting amount
+        uint32 afterSlash;
+        if (
+            exitingProvers[prover].exitRequestedAt != 0
+                && exitingProvers[prover].exitRequestedAt + EXIT_PERIOD < block.timestamp
+        ) {
+            // Exiting prover get's slashed
+            uint32 currentStaked = exitingProvers[prover].exitAmount;
+
+            afterSlash = uint32(
+                currentStaked * uint256(10_000 - SLASH_AMOUNT_IN_BP) / 10_000
+            );
+
+            exitingProvers[prover].exitAmount = afterSlash;
+        }
+        else if(topProvers[proverToId[prover]].rewardPerGas != 0) {
+            // Slash top prover
+            uint32 currentStaked = topProvers[proverToId[prover]].amount;
+
+            afterSlash = uint32(
+                currentStaked * uint256(10_000 - SLASH_AMOUNT_IN_BP) / 10_000
+            );
+
+            topProvers[proverToId[prover]].amount = afterSlash;
+        }
+
+        emit ProverSlashed(prover, afterSlash);
+    }
+
+    // @Daniel's comment: Adjust the staking. Users can use this funciton to
+    // stake, re-stake, exit,
+    // and change parameters.
+    // @Dani: Most prob. not for exit and modify, because it would require extra
+    // logic and cases. Let handle
+    // modification in a separate function
+    function stake(
+        uint32 totalAmount,
+        uint16 rewardPerGas,
+        uint16 capacity
+    )
+        external
+        nonReentrant
+    {
+
+        address newProver = _msgSender();
+        if(totalAmount == 0) {
```

Comment:
totalAmount is 0 is an full exit, but if totalAmount is 1. then it's still an exit. The code should handle them almost the same...